### PR TITLE
KAFKA-5017: Record batch first offset remains accurate after compaction

### PR DESCRIPTION
**Pull request by hachikuji**

### CODE REVIEW

File: clients/src/test/java/org/apache/kafka/common/record/MemoryRecordsTest.java [hunk 1]
```diff
@@ -302,38 +302,29 @@ protected boolean shouldRetain(RecordBatch recordBatch, Record record) {
     }
 
     @Test
-    public void testFilterToPreservesProducerInfo() {
+    public void testFilterToPreservesIdempotentProducerInfo() {
         if (magic >= RecordBatch.MAGIC_VALUE_V2) {
             ByteBuffer buffer = ByteBuffer.allocate(2048);
 
-            // non-idempotent, non-transactional
-            MemoryRecordsBuilder builder = MemoryRecords.builder(buffer, magic, compression, TimestampType.CREATE_TIME, 0L);
```
Comment (ijuma) [inline clients/src/test/java/org/apache/kafka/common/record/MemoryRecordsTest.java:310]: Sounds like the idea was to have have a mix of non-idempotent, idempotent and transactional batches together. I guess you think that there's no value in testing that? The benefit of separating them is clear, just making sure we are considering the potential drawback too.

**Review by ijuma — COMMENTED** [review]
Thanks for the PR, looks good overall. I left one comment and filed a PR with a few improvements:

https://github.com/hachikuji/kafka/pull/3

**Comment by hachikuji** [comment]
Let me summarize the changes in this patch. The main thing is a change in the compaction logic to _not_ preserve the base offset of each record batch. Previously if a record batch contained offsets [1,2,3] and if 1 was removed by the cleaner, the batch would still indicate 1 as its base offset. I couldn't think of a good reason to preserve the base offset from the original batch (we only need to ensure the last sequence is preserved for EoS) and it seemed a bit inconsistent with some of the actual usage. I therefore changed the behavior so that the new base offset always pointed to the offset of the first message (so it would be 2 using the example above). To reflect this change, I changed the name from "base offset" to "first offset". Similarly, I changed "base sequence" to "first sequence" and "base timestamp" to "first timestamp.

**Comment by hachikuji** [comment]
Going to close this since @junrao correctly pointed out that we must preserve the base sequence for deduplication. I will open a separate PR to add some comments on this point and fix some documentation inconsistencies.

**Pull request closed without merging**